### ISSUE TITLE
Integrate sub test (fixtures) output with outer test pretty printing

### ISSUE THREAD

**Issue by lucastheisen** [bats-core/bats-core#398]
Right now, if you follow the [fixture approach](https://github.com/bats-core/bats-core/blob/master/test/suite.bats#L12-L17) that bats itself uses, the output from failed subtests shows in tap format even when pretty printing.  For example, if i run:

outer.bats
```bash
#!/usr/bin/env bats

function error_exit {
  >&2 echo -e "$1"
  exit "${2:-1}"
}

function outer_runs_inner_one { #@test
  export INNER_ID=one
  run bats inner.bats
  if [[ "${status}" != "0" ]]; then
    error_exit "${output}"
  fi
}

function outer_runs_inner_two { #@test
  export INNER_ID=two
  run bats inner.bats
  if [[ "${status}" != "0" ]]; then
    error_exit "${output}"
  fi
}
```

inner.bats
```bash
#!/usr/bin/env bats

function inner_has_inner_id { #@test
  [[ -v INNER_ID ]]
}

function inner_inner_id_is_not_empty { #@test
  [[ -n "${INNER_ID}" ]]
}

function inner_inner_id_is_two { #@test
  [[ "${INNER_ID}" == "two" ]]
}
```

I get this output:
```console
ltheisen@MM233009-PC:/tmp/batssub$ bats outer.bats
 ✗ outer_runs_inner_one
   (from function `error_exit' in file outer.bats, line 5,
    in test file outer.bats, line 12)
     `error_exit "${status}: ${output}"' failed
   1: 1..3
   ok 1 inner_has_inner_id
   ok 2 inner_inner_id_is_not_empty
   not ok 3 inner_inner_id_is_two
   # (in test file inner.bats, line 12)
   #   `[[ "${INNER_ID}" == "two" ]]' failed
 ✓ outer_runs_inner_two

2 tests, 1 failure
```

It would be nice if the pretty printing were managed at the outer layer:

```console
ltheisen@MM233009-PC:/tmp/batssub$ bats outer.bats
 ✗ outer_runs_inner_one
   (from function `error_exit' in file outer.bats, line 5,
    in test file outer.bats, line 12)
     `error_exit "${output}"' failed
   ✓  inner_has_inner_id
   ✓  inner_inner_id_is_not_empty
   ✗  inner_inner_id_is_two
      # (in test file inner.bats, line 12)
      #   `[[ "${INNER_ID}" == "two" ]]' failed
 ✓ outer_runs_inner_two

2 tests, 1 failure
```

**Comment by martin-schulze-vireso** [comment]
What do you mean by "managed at the outer layer"?
Should bats detect that you `run bats` and format `$output` according to the formatter you chose for the outer bats run?

If you just want to enforce pretty printing, why not `run bats --pretty inner.bats`?

**Comment by lucastheisen** [comment]
> What do you mean by "managed at the outer layer"?
> Should bats detect that you run bats and format $output according to the formatter you chose for the outer bats run?

I could think of 2 different ways, either a `run_bats`/`bats_subtest`/`subtest` peer function of run that knows they are sub-tests and handles accordingly.  Or, a more generic, `--is-tap-output` style argument or env var that run could use to know that the output from what is being run is actually TAP output itself (and thus subtest).

As it stands, running with `run bats --pretty inner.bats` results in _unfriendly_ coloring and not indented to show the _sub_ nature of the tests:

> If you just want to enforce pretty printing, why not run bats --pretty inner.bats?

![image](https://user-images.githubusercontent.com/1952786/103420974-f63f7280-4b67-11eb-9968-8238e233732f.png)

In any case, i think it would be a good and valid use case to support some well defined approach to subtesting that would allow for consistently output and possibly even integrated test counts (sum up subtest to combine with overall count) or something.  In a perfect world, this approach would also allow that subtest _suite_ to have its results printed in realtime much like the parent test is now (as opposed to waiting for all subtests to complete before printing parent test result)...

This is definitely a _feature_ request, not any sort of bug or anything, but I would love to see it.

**Comment by martin-schulze-vireso** [comment]
Just to make sure there is no misunderstanding. You want a sub test feature which is intended for calling bats from a `.bats` file and has the following output features:

1. pretty format
2. the coloring of the outer test failure overrides the inner test's coloring
3. the indentation of the outer test is added to the inner tests'

However, my main question is about the "sub test" idea itself. IMHO, in the bats test suite, these fixtures are merely part of the setup for acceptance tests of bats itself. So the recursive nature comes through bats testing itself. In your case I don't see this as a primary feature.

What I see in your example is:

1. grouping of tests
2. parametrizing the tests from outside via an environment variable

I am wondering if these are the actual features you are looking for and we have an [xy problem](https://en.wikipedia.org/wiki/XY_problem) here?

**Comment by lucastheisen** [comment]
You are correct in your assumption that what I really need is parametrization.  The example in the initial text of this issue is my exact use case: outer runs all the tests of inner for multiple scenarios after setting different environment variables.

The subtest fixture idiom that's bats uses for it's own test is sufficient to my user case with the exception that the outputting of results is less than ideal for a few reasons:

* Results get buffered until the or test is complete making larger groups of tests appear to hang
* Default output switches to TAP which makes it much more difficult to debug failures when you have to sift through all the successful output as well
* Forcing subtest to use `--pretty` formatting gets confused with parent test formatting.

**Comment by martin-schulze-vireso** [comment]
I am thinking about file parametrization in #349 as well, feel free to add your input.

Regarding general TAP wrapping: while it sure would be an additional benefit, I fear that this niche use case might not be worth the added complexity.
